sno/init.py

```diff
@@ -43,200 +57,368 @@ def import_gpkg(ctx, geopackage, table, list_tables):
-    def __init__(self, prefixes=("GPKG",), suffix_required=False, **kwargs):
-        params = {
-            "exists": True,
-            "file_okay": True,
-            "dir_okay": False,
-            "writable": False,
-            "readable": True,
-            "resolve_path": False,
-            "allow_dash": False,
-            "path_type": None,
-        }
-        params.update(kwargs)
-
-        super().__init__(**params)
-
-        self.prefixes = prefixes
-        self.suffix_required = suffix_required
-
-    def convert(self, value, param, ctx):
-        if ":" not in value:
-            self.fail(f'expecting a prefix (eg. "GPKG:my.gpkg")')
-        prefix, value = value.split(":", 1)
+class OgrImporter:
+    """
+    Imports from an OGR source, currently from a whitelist of formats.
+    """
 
-        # need to deal with "GPKG:D:\foo\bar.gpkg:table"
-        search_from = 2 if is_windows else 0
+    @classmethod
+    def adapt_source_for_ogr(cls, source):
+        # Accept Path objects
+        ogr_source = str(source)
+        # Optionally, accept driver-prefixed paths like 'GPKG:'
+        allowed_formats = sorted(FORMAT_TO_OGR_MAP.keys())
+        m = re.match(
+            rf'^({"|".join(FORMAT_TO_OGR_MAP.keys())}):(.+)$', ogr_source, re.I
+        )
+        prefix = None
+        if m:
+            prefix, ogr_source = m.groups()
+            prefix = prefix.upper()
+            allowed_formats = [prefix]
+
+            if prefix in LOCAL_PATH_FORMATS:
+                # resolve GPKG:~/foo.gpkg and GPKG:~me/foo.gpkg
+                # usually this is handled by the shell, but the GPKG: prefix prevents that
+                ogr_source = os.path.expanduser(ogr_source)
+
+            if prefix in ('CSV', 'PG'):
+                # OGR actually handles these prefixes itself...
+                ogr_source = f'{prefix}:{ogr_source}'
+
+        if prefix in LOCAL_PATH_FORMATS:
+            if not os.path.exists(ogr_source):
+                raise NotFound(
+                    f"Couldn't find {ogr_source!r}", exit_code=NO_IMPORT_SOURCE
+                )
+        return ogr_source, allowed_formats
 
-        if ":" not in value[search_from:]:
-            if self.suffix_required:
-                self.fail(f'expecting a suffix (eg. "GPKG:my.gpkg:mytable")')
-            else:
-                suffix = None
-        else:
-            value, suffix = value.rsplit(":", 1)
+    @classmethod
+    def open(cls, source, table=None):
+        ogr_source, allowed_formats = cls.adapt_source_for_ogr(source)
+        allowed_ogr_drivers = [FORMAT_TO_OGR_MAP[x] for x in allowed_formats]
 
-        prefix = prefix.upper()
-        if prefix not in self.prefixes:
-            self.fail(
-                f'invalid prefix: "{prefix}" (choose from {", ".join(self.prefixes)})',
-                exit_code=INVALID_ARGUMENT,
+        writable = True
+        try:
+            # Try to open writable
+            ds = gdal.OpenEx(
```

Comment:
Why do we ever want a writeable import source? Seems... bad.

Reply:
Performance. It means we can use temp tables, in the case of postgres/GPKG (well, GPKG is overridden anyway, but the `iter_features_sorted` implementation uses temp tables)

Reply:
Only used it for `iter_features_sorted`, which I've now removed (#87 ) so I've removed this now.